### ISSUE TITLE
Unreliable BufferedUarte read_exact calls

### ISSUE THREAD

**Issue by huntc** [embassy-rs/embassy#536]
I'm having an issue receiving 18 bytes via `BufferedUarte` where not all of the bytes transmitted to it by another device are being received. I have used a scope to confirm that the other device is sending the 18 bytes.

I think it is to do with both the BufferedUarte and also how I read from it in my program. I read in two parts. First, we use `read_exact` to read a 5-byte header with the 5th byte indicating the remaining payload length. Then we `read_exact` the remainder payload.

If I change things to just one `read_exact` of 18 bytes instead of the two calls outlined above then all appears well.

At first, I thought my program might be exceeding the idle time logic (20 bit/baudrate timeout implemented via PPI), but disabling that and ensuring that the rx buffer was sized exactly to 18 bytes did not appear to help.

Another observation I made is that setting the size of the receive buffer to the exact size expected to be received makes the problem surface faster. If the buffer is large, then you might not see problems for a while.

**Comment by huntc** [comment]
I was able to reproduce this given a modification to the examples/nrf/buffered_uart.rs.

```rust
#![no_std]
#![no_main]
#![feature(type_alias_impl_trait)]

#[path = "../example_common.rs"]
mod example_common;

use embassy::executor::Spawner;
use embassy::io::{AsyncBufReadExt, AsyncWriteExt};
use embassy_nrf::buffered_uarte::State;
use embassy_nrf::gpio::NoPin;
use embassy_nrf::{buffered_uarte::BufferedUarte, interrupt, uarte, Peripherals};
use example_common::*;
use futures::pin_mut;

#[embassy::main]
async fn main(_spawner: Spawner, p: Peripherals) {
    let mut config = uarte::Config::default();
    config.parity = uarte::Parity::EXCLUDED;
    config.baudrate = uarte::Baudrate::BAUD115200;

    let mut tx_buffer = [0u8; 18];
    let mut rx_buffer = [0u8; 18]; // NOTE if I set this to a larger size then the problem is deferred!!!

    let irq = interrupt::take!(UARTE0_UART0);
    let mut state = State::new();
    let u = BufferedUarte::new(
        &mut state,
        p.UARTE0,
        p.TIMER0,
        p.PPI_CH0,
        p.PPI_CH1,
        irq,
        p.P0_08,
        p.P0_06,
        NoPin,
        NoPin,
        config,
        &mut rx_buffer,
        &mut tx_buffer,
    );
    pin_mut!(u);

    info!("uarte initialized!");

    // unwrap!(u.write_all(b"Hello!\r\n").await);
    // info!("wrote hello in uart!");

    // Simple demo, reading 8-char chunks and echoing them back reversed.
    loop {
        info!("reading...");
        let mut buf = [0u8; 18];
        unwrap!(u.read_exact(&mut buf[..5]).await);
        unwrap!(u.read_exact(&mut buf[5..]).await);
        info!("read done, got {}", buf);

        // Reverse buf
        // for i in 0..4 {
        //     buf.swap(i, 7 - i);
        // }

        // info!("writing...");
        // unwrap!(u.write_all(&buf).await);
        // info!("write done");

        // // Wait until the bytes are actually finished being transmitted
        // unwrap!(u.flush().await);
    }
}
```
Here's the first read of 18 bytes in two goes, which works just fine.

```
0 TRACE irq: start
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
1 TRACE   irq_rx: in state idle
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
2 TRACE   irq_rx: starting 18
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
3 TRACE   irq_rx: buf 536871139 18
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
4 TRACE   irq_tx: in state Idle
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
5 TRACE irq: end
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
6 INFO  uarte initialized!
└─ buffered_uart::__embassy_main::task::{generator#0} @ src/bin/buffered_uart.rs:44
7 INFO  reading...
└─ buffered_uart::__embassy_main::task::{generator#0} @ src/bin/buffered_uart.rs:51
8 TRACE poll_read
└─ embassy_nrf::buffered_uarte::{impl#3}::poll_fill_buf::{closure#0} @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
9 TRACE   empty
└─ embassy_nrf::buffered_uarte::{impl#3}::poll_fill_buf::{closure#0} @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
10 TRACE irq: start
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
11 TRACE   irq_rx: in state receiving
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
12 TRACE   irq_rx: endrx 18
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
13 TRACE   irq_rx: in state idle
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
14 TRACE   irq_tx: in state Idle
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
15 TRACE irq: end
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
16 TRACE poll_read
└─ embassy_nrf::buffered_uarte::{impl#3}::poll_fill_buf::{closure#0} @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
17 TRACE   got 536871139 18
└─ embassy_nrf::buffered_uarte::{impl#3}::poll_fill_buf::{closure#0} @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
18 TRACE consume 5
└─ embassy_nrf::buffered_uarte::{impl#3}::consume::{closure#0} @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
19 TRACE irq: start
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
20 TRACE   irq_rx: in state idle
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
21 TRACE   irq_rx: starting 5
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
22 TRACE   irq_rx: buf 536871139 5
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
23 TRACE   irq_tx: in state Idle
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
24 TRACE irq: end
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
25 TRACE poll_read
└─ embassy_nrf::buffered_uarte::{impl#3}::poll_fill_buf::{closure#0} @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
26 TRACE   got 536871144 13
└─ embassy_nrf::buffered_uarte::{impl#3}::poll_fill_buf::{closure#0} @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
27 TRACE consume 13
└─ embassy_nrf::buffered_uarte::{impl#3}::consume::{closure#0} @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
28 TRACE irq: start
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
29 TRACE   irq_rx: in state receiving
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
30 TRACE   irq_tx: in state Idle
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
31 TRACE irq: end
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
32 INFO  `read done,` got [48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 48, 49, 50, 51, 52, 53, 54, 55]
└─ buffered_uart::__embassy_main::task::{generator#0} @ src/bin/buffered_uart.rs:55
```

Here's the continuation from the second time around that where we attempt to read another 18 bytes. You'll note that we don't get the `read done` as per line 32 above. Line 38 shows endrx as 5, and line 48 shows endrx as 6. There are no further endrx logs.

```
33 INFO  reading...
└─ buffered_uart::__embassy_main::task::{generator#0} @ src/bin/buffered_uart.rs:51
34 TRACE poll_read
└─ embassy_nrf::buffered_uarte::{impl#3}::poll_fill_buf::{closure#0} @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
35 TRACE   empty
└─ embassy_nrf::buffered_uarte::{impl#3}::poll_fill_buf::{closure#0} @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
36 TRACE irq: start
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
37 TRACE   irq_rx: in state receiving
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
38 TRACE   irq_rx: endrx 5
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
39 TRACE   irq_rx: in state idle
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
40 TRACE   irq_rx: starting 13
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
41 TRACE   irq_rx: buf 536871144 13
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
42 TRACE   irq_tx: in state Idle
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
43 TRACE irq: end
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
44 TRACE poll_read
└─ embassy_nrf::buffered_uarte::{impl#3}::poll_fill_buf::{closure#0} @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
45 TRACE   got 536871139 5
└─ embassy_nrf::buffered_uarte::{impl#3}::poll_fill_buf::{closure#0} @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
46 TRACE irq: start
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
47 TRACE   irq_rx: in state receiving
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
48 TRACE   irq_rx: endrx 6
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
49 TRACE   irq_rx: in state idle
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
50 TRACE   irq_rx: starting 7
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
51 TRACE   irq_rx: buf 536871150 7
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
52 TRACE   irq_tx: in state Idle
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
53 TRACE irq: end
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
54 TRACE consume 5
└─ embassy_nrf::buffered_uarte::{impl#3}::consume::{closure#0} @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
55 TRACE irq: start
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
56 TRACE   irq_rx: in state receiving
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
57 TRACE   irq_tx: in state Idle
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
58 TRACE irq: end
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
59 TRACE poll_read
└─ embassy_nrf::buffered_uarte::{impl#3}::poll_fill_buf::{closure#0} @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
60 TRACE   got 536871144 6
└─ embassy_nrf::buffered_uarte::{impl#3}::poll_fill_buf::{closure#0} @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
61 TRACE consume 6
└─ embassy_nrf::buffered_uarte::{impl#3}::consume::{closure#0} @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
62 TRACE irq: start
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
63 TRACE   irq_rx: in state receiving
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
64 TRACE   irq_tx: in state Idle
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
65 TRACE irq: end
└─ embassy_nrf::buffered_uarte::{impl#6}::on_interrupt @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
66 TRACE poll_read
└─ embassy_nrf::buffered_uarte::{impl#3}::poll_fill_buf::{closure#0} @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
67 TRACE   empty
└─ embassy_nrf::buffered_uarte::{impl#3}::poll_fill_buf::{closure#0} @ /Users/huntc/Projects/hacking/embassy/embassy-nrf/src/fmt.rs:112
```

**Comment by huntc** [comment]
If we modify the program to:

```rust
        unwrap!(u.read_exact(&mut buf[..18]).await);
        // unwrap!(u.read_exact(&mut buf[5..]).await);
```

...then all is well!

**Comment by huntc** [comment]
Further to the matrix chats, it turns out that the functionality here is presently reliable only when hardware flow control is used. This is a problem that we should address at some stage.

**Comment by Dirbaio** [comment]
I believe this should be solved by #537
